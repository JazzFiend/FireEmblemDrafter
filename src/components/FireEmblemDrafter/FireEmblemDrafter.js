import React, { Component } from "react";
import { hot } from "react-hot-loader";
import _ from 'lodash';
import PropTypes from 'prop-types';

import './FireEmblemDrafter.css';
import Draft from '../../logic/Draft';
import RandomNumberGenerator from '../../logic/RandomNumberGenerator';
import RandomElementSelector from '../../logic/RandomElementSelector';
import TestRandomNumberGenerator from '../../logic/__mocks__/TestRandomNumberGenerator';
import { Pick } from "../Pick";

export class FireEmblemDrafter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pick: Array(6).fill(undefined),
      draftInProgress: false,
      team: Array(0),
      draft: undefined,
      roster: props.roster,
      teamSize: props.teamSize,
      randomizePicks: props.isRandom,
    }
  }

  render() {
    return (
      <div>
        <button className="start-draft-button" data-testid="start-draft-button" onClick={() => this.startDraft()}>
          {this.state.draftInProgress ? 'Restart Draft!' : 'Start Draft!'}
        </button>
        {this.showPick()}
        <div data-testid='team-list'>
          {this.displayTeam(this.state.team)}
        </div>
      </div>
    );
  }

  startDraft() {
    let draft;
    if(this.state.randomizePicks) {
      draft = new Draft(this.state.roster, this.state.teamSize, new RandomElementSelector(new RandomNumberGenerator()));
    } else {
      draft = new Draft(this.state.roster, this.state.teamSize, new RandomElementSelector(new TestRandomNumberGenerator()));
    }

    this.setState({
      draftInProgress: true,
      draft: draft,
      team: Array(0),
    });

    this.generatePick(draft);
  }

  generatePick(draft) {
    let pick = draft.generateNextPick();
    this.setState({
      pick: pick,
    });
  }

  showPick() {
    return (
      <div>
        {
          this.state.draftInProgress && 
          <Pick pick={this.state.pick} onClick={index => this.handlePickClick(index)}/>
        }
      </div>
    );
  }

  handlePickClick(buttonIndex) {
    let draft = this.state.draft;
    let pick = this.state.pick;

    this.selectPick(pick[buttonIndex]);
    if(draft.isDraftFinished()) {
      this.completeDraft();
    } else {
      this.generatePick(draft);
    }
  }

  selectPick(pickIndex) {
    this.state.draft.select(pickIndex);
    this.setState({
      pick: Array(6).fill(undefined),
      team: this.state.draft.getCurrentTeam(),
    });
  }

  completeDraft() {
    this.setState({
      draftInProgress: false,
      draft: undefined,
    });
  }

  displayTeam(team) {
    let teamDisplayFriendly = "";
    _.forEach(team, function(value) {
      teamDisplayFriendly = teamDisplayFriendly + value + '\n';
    });
    return teamDisplayFriendly;
  }
}

export default hot(module)(FireEmblemDrafter);

FireEmblemDrafter.propTypes = {
  roster: PropTypes.arrayOf(PropTypes.string),
  teamSize: PropTypes.number,
  isRandom: PropTypes.bool,
};
