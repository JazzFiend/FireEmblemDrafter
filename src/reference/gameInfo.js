const gameInfo = [
  {
    id: 0,
    title: 'The Sacred Stones',
    playableCharacters: [
      'Eirika', 'Ephraim', 'Seth', 'Franz', 'Gilliam', 'Vanessa', 'Moulder', 'Ross', 'Garcia', 'Neimi', 'Colm',
      'Artur', 'Lute', 'Natasha', 'Joshua', 'Forde', 'Kyle', 'Tana', 'Amelia', 'Innes', 'Gerik', 'Tethys', 'Marisa',
      'Ewan', 'Duessel', 'Cormag', 'L\'Arachel', 'Dozla', 'Saleh', 'Rennac', 'Knoll', 'Myrrh', 'Syrene',
    ],
    exclusiveCharacters: [],
    defaultTeamSize: 18,
  },
  {
    id: 1,
    title: 'The Blazing Blade',
    playableCharacters: [
      'Lyn', 'Sain', 'Kent', 'Florina', 'Wil', 'Dorcas', 'Serra', 'Erk', 'Rath', 'Matthew', 'Lucius', 'Wallace',
      'Eliwood', 'Marcus', 'Lowen', 'Rebecca', 'Bartre', 'Hector', 'Oswin', 'Guy', 'Priscilla', 'Raven', 'Canas',
      'Dart', 'Fiora', 'Legault', 'Ninian', 'Isadora', 'Heath', 'Hawkeye', 'Geitz', 'Pent', 'Louise', 'Karel',
      'Harken', 'Nino', 'Jaffar', 'Vaida', 'Renault', 'Farina', 'Karla',
    ],
    exclusiveCharacters: [
      ['Wallace', 'Geitz'],
      ['Karel', 'Harken'],
    ],
    defaultTeamSize: 15,
  },
  {
    id: 2,
    title: 'The Binding Blade',
    playableCharacters: [
      'Roy', 'Marcus', 'Alen', 'Lance', 'Wolt', 'Bors', 'Merlinus', 'Elen', 'Dieck', 'Wade', 'Lot', 'Shanna', 'Chad',
      'Lugh', 'Clarine', 'Rutger', 'Saul', 'Dorothy', 'Sue', 'Zelot', 'Trec', 'Noah', 'Astolfo', 'Lilina', 'Gwendolyn',
      'Barthe', 'Ogier', 'Fir', 'Sin', 'Gonzalez', 'Geese', 'Klein', 'Thea', 'Larum', 'Echidna', 'Elffin', 'Bartre',
      'Raigh', 'Cath', 'Melady', 'Perceval', 'Cecilia', 'Sophia', 'Igrene', 'Garret', 'Fae', 'Hugh', 'Zeiss',
      'Douglas', 'Niime', 'Juno', 'Dayan', 'Yoder', 'Karel',
    ],
    exclusiveCharacters: [
      ['Bartre', 'Echidna'],
      ['Larum', 'Elffin'],
      ['Juno', 'Dayan'],
    ],
    defaultTeamSize: 18,
  },
];

export default gameInfo;
