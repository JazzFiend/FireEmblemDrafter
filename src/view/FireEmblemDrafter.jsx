import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import GameSelectorController from '../controllers/GameSelectorController';
import DraftController from '../controllers/DraftController';
import RosterModifiersController from '../controllers/RosterModifiersController';
import DraftOptionsController from '../controllers/DraftOptionsController';

class FireEmblemDrafter extends PureComponent {
  showGameSelectorController() {
    const {
      draftInProgress, gameInfo, handleGameSelector,
    } = this.props;

    return (
      <GameSelectorController
        draftInProgress={draftInProgress}
        gameInfo={gameInfo}
        handleGameSelector={handleGameSelector}
      />
    );
  }

  showRosterController() {
    const {
      roster,
      restrictedCharacters,
      requiredCharacters,
      draftInProgress,
      handleRestrictedUnitSelector,
      handleRequiredUnitSelector,
    } = this.props;

    return (
      <div>
        <RosterModifiersController
          restrictedCharacters={restrictedCharacters}
          requiredCharacters={requiredCharacters}
          allCharacters={roster}
          handleRestrictedUnitSelector={handleRestrictedUnitSelector}
          handleRequiredUnitSelector={handleRequiredUnitSelector}
          draftInProgress={draftInProgress}
        />
      </div>
    );
  }

  showDraftOptionsController() {
    const {
      gameSelected,
      draftInProgress,
      handleTeamSizeChange,
      teamSize,
      maxTeamSize,
    } = this.props;

    return (
      <DraftOptionsController
        gameSelected={gameSelected}
        draftInProgress={draftInProgress}
        handleTeamSizeChange={handleTeamSizeChange}
        defaultValue={teamSize}
        maxTeamSize={maxTeamSize}
      />
    );
  }

  showDraftController() {
    const {
      draftInProgress,
      roster,
      restrictedCharacters,
      requiredCharacters,
      teamSize,
      randomizePicks,
      handleDraftProgress,
    } = this.props;

    return (
      <DraftController
        draftInProgress={draftInProgress}
        handleDraftProgress={handleDraftProgress}
        roster={roster}
        restrictedCharacters={restrictedCharacters}
        requiredCharacters={requiredCharacters}
        teamSize={teamSize}
        randomizePicks={randomizePicks}
      />
    );
  }

  render() {
    return (
      <div>
        {this.showGameSelectorController()}
        {this.showRosterController()}
        {this.showDraftOptionsController()}
        {this.showDraftController()}
      </div>
    );
  }
}
export default FireEmblemDrafter;

FireEmblemDrafter.propTypes = {
  draftInProgress: PropTypes.bool,
  gameSelected: PropTypes.bool.isRequired,
  randomizePicks: PropTypes.bool,
  teamSize: PropTypes.number,
  maxTeamSize: PropTypes.number.isRequired,
  roster: PropTypes.arrayOf(PropTypes.string),
  requiredCharacters: PropTypes.arrayOf(PropTypes.string),
  restrictedCharacters: PropTypes.arrayOf(PropTypes.string),
  gameInfo: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
  })).isRequired,
  handleDraftProgress: PropTypes.func,
  handleGameSelector: PropTypes.func,
  handleRequiredUnitSelector: PropTypes.func,
  handleRestrictedUnitSelector: PropTypes.func,
  handleTeamSizeChange: PropTypes.func,
};

FireEmblemDrafter.defaultProps = {
  draftInProgress: false,
  roster: [],
  restrictedCharacters: [],
  requiredCharacters: [],
  teamSize: 0,
  randomizePicks: false,
  handleGameSelector: () => {},
  handleDraftProgress: () => {},
  handleRestrictedUnitSelector: () => {},
  handleRequiredUnitSelector: () => {},
  handleTeamSizeChange: () => {},
};
