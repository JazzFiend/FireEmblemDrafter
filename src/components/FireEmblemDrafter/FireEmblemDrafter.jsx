import React, { Component } from "react";
import { hot } from "react-hot-loader";
import _ from 'lodash';
import PropTypes from 'prop-types';

import './FireEmblemDrafter.css';
import Draft from '../../logic/Draft';
import RandomNumberGenerator from '../../logic/RandomNumberGenerator';
import RandomElementSelector from '../../logic/RandomElementSelector';
import TestRandomNumberGenerator from '../../logic/__mocks__/TestRandomNumberGenerator';
import Pick from '../Pick';

export class FireEmblemDrafter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pick: Array(6).fill(undefined),
      draftInProgress: false,
      team: Array(0),
      draft: undefined,
      roster: props.roster,
      teamSize: props.teamSize,
      randomizePicks: props.isRandom,
    }
  }

  startDraft(roster, teamSize, randomizePicks) {
    let randomizer = this.createRandomizer(randomizePicks);
    let draft = new Draft(roster, teamSize, randomizer);
    let pick = draft.generateNextPick();

    this.setState({
      draftInProgress: true,
      draft: draft,
      team: Array(0),
      pick: pick,
    });
  }

  createRandomizer(randomizePicks) {
    if (randomizePicks) {
      return new RandomElementSelector(new RandomNumberGenerator());
    }
    else {
      return new RandomElementSelector(new TestRandomNumberGenerator());
    }
  }

  showPick(pick, draftInProgress, draft) {
    return (
      <div>
        {
          draftInProgress && 
          <Pick pick={pick} onClick={index => this.handlePickClick(index, pick, draft)} />
        }
      </div>
    );
  }

  handlePickClick(buttonIndex, pick, draft) {
    let newPick;

    draft.select(pick[buttonIndex]);
    if(!draft.isDraftFinished()) {
      newPick = draft.generateNextPick();
    }

    this.setState({
      draftInProgress: !(draft.isDraftFinished()),
      pick: newPick,
      team: draft.getCurrentTeam(),
    });
  }

  displayTeam(team) {
    let teamDisplayFriendly = "";
    _.forEach(team, function(value) {
      teamDisplayFriendly = teamDisplayFriendly + value + '\n';
    });
    return teamDisplayFriendly;
  }

  render() {
    const stateCopy = this.state;
    return (
      <div>
        <button type="button" className="start-draft-button" data-testid="start-draft-button" onClick={() => this.startDraft(stateCopy.roster, stateCopy.teamSize, stateCopy.randomizePicks)}>
          {stateCopy.draftInProgress ? 'Restart Draft!' : 'Start Draft!'}
        </button>
        {this.showPick(stateCopy.pick, stateCopy.draftInProgress, stateCopy.draft)}
        <div data-testid='team-list'>
          {this.displayTeam(stateCopy.team)}
        </div>
      </div>
    );
  }
}
export default hot(module)(FireEmblemDrafter);

FireEmblemDrafter.propTypes = {
  roster: PropTypes.arrayOf(PropTypes.string),
  teamSize: PropTypes.number,
  isRandom: PropTypes.bool,
};

FireEmblemDrafter.defaultProps = {
  roster: [],
  teamSize: 0,
  isRandom: false,
};
