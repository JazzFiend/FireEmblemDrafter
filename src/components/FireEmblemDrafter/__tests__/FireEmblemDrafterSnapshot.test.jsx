import React from 'react';
import {
  render,
  cleanup,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { FireEmblemDrafter } from '../FireEmblemDrafter';
import gameInfo from '../../../reference/gameInfo';

jest.mock('../../GameSelector', () => () => 'GameSelector');
jest.mock('../../DraftController', () => () => 'DraftController');

afterEach(cleanup);

test('renders and matches snapshot', () => {
  const { asFragment } = render(
    <FireEmblemDrafter
      teamSize={2}
      isRandom={false}
      gameInfo={gameInfo}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});
