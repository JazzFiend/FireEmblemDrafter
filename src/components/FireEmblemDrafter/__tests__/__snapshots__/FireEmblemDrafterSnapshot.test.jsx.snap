// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    GameSelector
    <div>
      <div>
        <div
          data-testid="required-drop-down"
        />
        <div
          data-testid="required-list"
        />
        <div
          data-testid="restricted-drop-down"
        />
        <div
          data-testid="restricted-list"
        />
      </div>
    </div>
    DraftController
  </div>
</DocumentFragment>
`;
